text/0000-http-index.md

```diff
@@ -0,0 +1,139 @@
+
+## Greedy fetch
+
+To simplify the implementation, and parallelize fetches effectively, Cargo will have to fetch all dependency information before performing the actual dependency resolution algorithm. This means it'll have to pessimistically fetch information about all sub dependencies of all dependency versions that *may* match known version requirements. This won't add much overhead, because requests are per create, not per crate version. It causes additional fetches only for dependencies that were used before, but were later dropped. Fetching is still narrowed by required version ranges, so even worst cases can be avoided by bumping version requirements. For example:
+
+* foo v1.0.1 depends on old-dep v1.0.0
+* foo v1.0.2 depends on maybe-dep v1.0.2
+* foo v1.0.3 depends on maybe-dep v1.0.3
+* foo v1.0.4 has no dependencies
+
+If a dependency requires `foo >=1.0.2`, then Cargo would need to fetch information about `maybe-dep` (once), even if `foo v1.0.4` ends up being selected later. However, it would not need to fetch `old-dep`. If the version requirement was upgraded to `foo >=v1.0.4` then there wouldn't be any extra fetches.
+
+## Offline support
+
+The proposed solution fully preserves Cargo's ability to work offline. Fetching of crates while online by necessity downloads enough of the index to use them, and all this data remains cached for use offline.
+
+## Bandwidth reduction
+
+Cargo supports HTTP/2, which handles many similar requests efficiently.
+
+All fetched dependency files can be cached, and refreshed using conditional HTTP requests (with `Etag` or `If-Modified-Since` headers), to avoid redownloading of files that haven't changed.
+
+Dependency files compress well. Currently the largest file of `rustc-ap-rustc_data_structures` compresses from 1MiB to 26KiB with Brotli. Many servers support transparently serving pre-compressed files (i.e. request for `/rustc-ap-rustc_data_structures` can be served from `rustc-ap-rustc_data_structures.gz` with an appropriate content encoding header), so the index can use high compression levels without increasing CPU cost of serving the files.
+
+Even in the worst case of downloading the entire index file by file, it should still use significantly less bandwidth than git clone (individually compressed files add up to about 39MiB).
+
+## Optionally, a rotated incremental changelog
+
+To further reduce number requests needed to update the index, the index may maintain an append-only log of changes. For each change (crate version published or yanked), the log would append a line with: epoch number (explained below), last-modified timestamp, and the name of the changed crate, e.g.
+
+```
+1 2019-10-18 23:51:23 oxigen
+1 2019-10-18 23:51:25 linda
+1 2019-10-18 23:51:29 rv
+1 2019-10-18 23:52:00 anyhow
+1 2019-10-18 23:53:03 build_id
+1 2019-10-18 23:56:16 canonical-form
+1 2019-10-18 23:59:01 cotton
+1 2019-10-19 00:01:44 kg-utils
+1 2019-10-19 00:08:45 serde_traitobject
```

Comment:
```suggestion
1 2019-10-18T23:51:23Z oxigen
1 2019-10-18T23:51:25Z linda
1 2019-10-18T23:51:29Z rv
1 2019-10-18T23:52:00Z anyhow
1 2019-10-18T23:53:03Z build_id
1 2019-10-18T23:56:16Z canonical-form
1 2019-10-18T23:59:01Z cotton
1 2019-10-19T00:01:44Z kg-utils
1 2019-10-19T00:08:45Z serde_traitobject
```

I guess it would probably be best to use ISO8601 dates in UTC here to simplify the parsing.